extra_data/keydata.py

```diff
@@ -232,8 +235,22 @@ def split_trains(self, parts=None, trains_per_part=None):
             A maximum number of trains in each part. Parts will often have
             fewer trains than this.
         """
-        for s in split_trains(len(self.train_ids), parts, trains_per_part):
-            yield self.select_trains(s)
+        # tids_files points to the file for each train.
+        # This avoids checking all files for each chunk, which can be slow.
+        tids_files = [None] * len(self.train_ids)
+        tid_to_ix = {t: i for i, t in enumerate(self.train_ids)}
+        for file in self.files:
+            f_tids = file.train_ids if self.inc_suspect_trains else file.valid_train_ids
+            for tid in f_tids:
+                ix = tid_to_ix.get(tid, None)
+                if ix is not None:
+                    tids_files[ix] = file
+
+        for sl in split_trains(len(self.train_ids), parts, trains_per_part):
+            tids = self.train_ids[sl]
+            files = set(tids_files[sl]) - {None}
+            files = sorted(files, key=lambda f: f.filename)
+            yield self._only_tids(tids, files=files)
```

Comment:
If I'm not mistaken this is exactly the same for `SourceData`? I'd say it's complicated enough to go into a helper function.

Reply:
Good call, yeah. How about we split out `trains_file_index(train_ids, files)` as a helpeer function?